src/main/java/org/broadinstitute/hellbender/utils/nio/NioFileCopier.java

```diff
@@ -0,0 +1,624 @@
+package org.broadinstitute.hellbender.utils.nio;
+
+import org.apache.commons.collections4.queue.CircularFifoQueue;
+import org.apache.commons.io.FileUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+
+/**
+ * Class to copy a file using {@link java.nio}.
+ * Operates using paths.
+ *
+ * INSTANCES OF THIS CLASS ARE NOT THREAD-SAFE!
+ *
+ * Created by jonn on 8/27/18.
+ */
+public class NioFileCopier {
+
+    //==================================================================================================================
+    // Standard logger:
+    private static final Logger logger = LogManager.getLogger(NioFileCopier.class);
+
+    //==================================================================================================================
+    // Public Static Members:
+
+    //==================================================================================================================
+    // Private Static Members:
+
+    private static final int    BUFFER_SIZE_BYTES                  = 1024 * 1024;
+    private static final double PROGRESS_DISPLAY_PERCENT_INCREMENT = 0.25;
+
+    private static final long SECOND_IN_MS = 1000;
+    private static final long MINUTE_IN_MS = SECOND_IN_MS * 60;
+    private static final long HOUR_IN_MS   = MINUTE_IN_MS * 60;
+    private static final long DAY_IN_MS    = HOUR_IN_MS * 24;
+
+    private static final int KB_TO_BYTES = 1024;
+    private static final int MS_TO_SEC = 1000;
+    private static final int NANOS_TO_MILLIS  = 1000000;
+    private static final int NANOS_TO_SECONDS = 1000000000;
+
+    private static final int COPY_SPEED_HISTORY_SIZE = 10;
+
+    private static final boolean OVERWRITE_EXISTING_DEFAULT = false;
+    private static final boolean DO_LOG_PROGRESS_DEFAULT    = true;
+    private static final boolean DO_LOG_VERBOSE_DEFAULT     = false;
+
+    //==================================================================================================================
+    // Private Members:
+
+    // Data variables:
+    protected final Path source;
+    protected final Path dest;
+
+    private long srcFileSize;
+    private int srcFileSizeNumDigits;
+
+    private Map<ChecksumCalculator, String> checksumMap              = new HashMap<>();
```

Comment:
The check sum map seems like an unnecessary complication.  Is there a use case for running multiple checksums and running check sums repeatedly?

Reply:
Not right now.  I'll remove the map and make it a single checksum.